src/plugins/validation/oas3/semantic-validators/pagination-ibm.js

```diff
@@ -0,0 +1,229 @@
+    );
+    if (startParamIndex !== -1) {
+      const startParam = params[startParamIndex];
+      if (
+        !startParam.schema ||
+        startParam.schema.type !== 'string' ||
+        !!startParam.required
+      ) {
+        result[checkStatus].push({
+          path: ['paths', path, 'get', 'parameters', startParamIndex],
+          message: `The ${
+            startParam.name
+          } parameter must be of type string and optional.`
+        });
+      }
+    }
+
+    // - The response body must contain a `limit` property that is type integer and required
+
+    const limitProp = jsonResponse.schema.properties.limit;
+    if (!limitProp) {
+      result[checkStatus].push({
+        path: [
+          'paths',
+          path,
+          'get',
+          'responses',
+          resp,
+          'content',
+          'application/json',
+          'schema',
+          'properties'
+        ],
+        message: `The response body of a paginated list operation must contain a "limit" property.`
+      });
+    } else if (
+      limitProp.type !== 'integer' ||
+      !jsonResponse.schema.required ||
+      jsonResponse.schema.required.indexOf('limit') === -1
+    ) {
+      result[checkStatus].push({
+        path: [
+          'paths',
+          path,
+          'get',
+          'responses',
+          resp,
+          'content',
+          'application/json',
+          'schema',
+          'properties',
+          'limit'
+        ],
+        message: `The "limit" property in the response body of a paginated list operation must be of type integer and required.`
+      });
+    }
+
+    // - If the operation has an `offset` query parameter, the response body must contain an `offset` property this is type integer and required
+
+    if (offsetParamIndex !== -1) {
+      const offsetProp = jsonResponse.schema.properties.offset;
+      if (!offsetProp) {
+        result[checkStatus].push({
+          path: [
+            'paths',
+            path,
+            'get',
+            'responses',
+            resp,
+            'content',
+            'application/json',
+            'schema',
+            'properties'
+          ],
+          message: `The response body of a paginated list operation must contain a "offset" property.`
+        });
+      } else if (
+        offsetProp.type !== 'integer' ||
+        !jsonResponse.schema.required ||
+        jsonResponse.schema.required.indexOf('offset') === -1
+      ) {
+        result[checkStatus].push({
+          path: [
+            'paths',
+            path,
+            'get',
+            'responses',
+            resp,
+            'content',
+            'application/json',
+            'schema',
+            'properties',
+            'offset'
+          ],
+          message: `The "offset" property in the response body of a paginated list operation must be of type integer and required.`
+        });
+      }
+    }
+  }
```

Comment:
Since these two checks look identical in every way but the name of the parameter, I would suggest creating a new function to handle this and passing in both "limit" and "offset".

Something like
```js
function doTheCheck(paramName) {
  const prop = jsonResponse.schema.properties[paramName];
  if (!prop) {
    result[checkStatus].push({
      path: [
        'paths',
        path,
        'get',
        'responses',
        resp,
        'content',
        'application/json',
        'schema',
        'properties'
      ],
      message: `The response body of a paginated list operation must contain a "${paramName}" property.`
    });
  } else if (
    prop.type !== 'integer' ||
    !jsonResponse.schema.required ||
    jsonResponse.schema.required.indexOf(paramName) === -1
  ) {
    result[checkStatus].push({
      path: [
        'paths',
        path,
        'get',
        'responses',
        resp,
        'content',
        'application/json',
        'schema',
        'properties',
        paramName
      ],
      message: `The "${paramName}" property in the response body of a paginated list operation must be of type integer and required.`
    });
  }
}
```

It would need access to variables like `jsonResponse` and `result` but I think you can make it a closure inside the `validate` method to maintain their context. This would remove some code, but it's not a big deal. So, up to you.

Reply:
Rather than create a function for this, I just created a variable to hold the common elements of the path.  By itself this eliminates almost 30 lines of code, but keeps the flexibility of the separate logic for these two parameters.

Reply:
That sounds good to me, my spidey-senses just get tingly when I see opportunity to reduce duplicated logic 🙂